plutus-benchmark/linear-vesting/src/LinearVesting.hs

```diff
@@ -0,0 +1,243 @@
+          traceError "Unlock not permitted until firstUnlockPossibleAfter time"
+      | newRemainingQty <= 0 ->
+          traceError "Zero remaining assets not allowed"
+      | newRemainingQty > oldRemainingQty ->
+          traceError "Remaining asset exceed old asset"
+      | expectedRemainingQty /= newRemainingQty ->
+          traceError "Mismatched remaining asset"
+      | txOutDatum resolvedOut /= outputDatum ->
+          traceError "Datum Modification Prohibited"
+      | countInputsAtScript scriptHash inputs /= 1 ->
+          traceError "Double satisfaction"
+      | otherwise ->
+          True
+
+validateVestingFullUnlock :: ScriptContext -> Bool
+validateVestingFullUnlock ctx =
+  let
+    txInfo :: TxInfo = scriptContextTxInfo ctx
+    currentTimeApproximation :: Integer =
+      getPOSIXTime (getLowerInclusiveTimeRange (txInfoValidRange txInfo))
+    SpendingScript _ownRef (Just (Datum datum)) = scriptContextScriptInfo ctx
+    vestingDatum :: VestingDatum = unsafeFromBuiltinData datum
+    PubKeyCredential beneficiaryKey = addressCredential (beneficiary vestingDatum)
+   in
+    if
+      | not (txSignedBy txInfo beneficiaryKey) ->
+          traceError "Missing beneficiary signature"
+      | vestingPeriodEnd vestingDatum >= currentTimeApproximation ->
+          traceError "Unlock not permitted until vestingPeriodEnd time"
+      | otherwise ->
+          True
+
+getLowerInclusiveTimeRange :: POSIXTimeRange -> POSIXTime
+getLowerInclusiveTimeRange = \case
+  Interval (LowerBound (Finite posixTime) inclusive) _upperBound ->
+    if inclusive then posixTime else posixTime + 1
+  _ -> traceError "Time range not Finite"
+
+{-# INLINEABLE typedValidator #-}
+typedValidator :: ScriptContext -> Bool
+typedValidator context =
+  trace "Validation completed"
+    $ case redeemer of
+      FullUnlock ->
+        validateVestingFullUnlock $ trace "Full unlock requested" context
+      PartialUnlock ->
+        validateVestingPartialUnlock $ trace "Partial unlock requested" context
+ where
+  {-# INLINEABLE redeemer #-}
+  redeemer :: VestingRedeemer
+  redeemer =
+    case fromBuiltinData (getRedeemer (scriptContextRedeemer context)) of
+      Nothing -> traceError "Failed to parse Redeemer"
+      Just r  -> trace "Parsed Redeemer" r
+
+{-# INLINEABLE untypedValidator #-}
+untypedValidator :: BuiltinData -> BuiltinUnit
+untypedValidator scriptContextData =
+  case trace "Parsing ScriptContext..." (fromBuiltinData scriptContextData) of
+    Nothing  -> traceError "Failed to parse ScriptContext"
+    Just ctx -> check $ typedValidator $ trace "Parsed ScriptContext" ctx
+
+validatorCode :: CompiledCode (BuiltinData -> BuiltinUnit)
+validatorCode = $$(compile [||untypedValidator||])
+
+validatorCodeFullyApplied :: CompiledCode BuiltinUnit
+validatorCodeFullyApplied =
+  validatorCode `unsafeApplyCode` liftCodeDef (toBuiltinData testScriptContext)
+
+testScriptContext :: ScriptContext
```

Comment:
Let's have this as an another file.

Reply:
Domain types can stay in `LinearVesting.hs` and the test cases file can import the domain types and the script itself. 

My argument is if we have it on same file, it will be more difficult to differentiate changes made in test cases and/or that in the script itself in the future. Basically for future tracking and review.

If you see this as something we add once here and rarely make changes or additions, I'm fine with having everything in one file.

Reply:
Done.